pkg/target/target_test.go

```diff
@@ -682,3 +694,75 @@ func TestMatcher_Match(t *testing.T) {
 		})
 	}
 }
+
+func TestNamespaceCache(t *testing.T) {
+	ns1 := makeNamespace("my-ns1", map[string]string{"ns1": "label"})
+	ns2 := makeNamespace("my-ns2", map[string]string{"ns2": "label"})
+	tests := []struct {
+		name    string
+		addNs   map[string]interface{}
+		wantNs  map[string]*corev1.Namespace
+		wantErr bool
+	}{
+		{
+			name:    "retrieving a namespace from empty cache returns nil",
+			addNs:   map[string]interface{}{},
+			wantNs:  map[string]*corev1.Namespace{"my-ns2": nil},
+			wantErr: false,
+		},
+		{
+			name:    "retrieving a namespace that doesnt exist returns nil",
+			addNs:   map[string]interface{}{"my-ns1": ns1},
+			wantNs:  map[string]*corev1.Namespace{"my-ns2": nil},
+			wantErr: false,
+		},
+		{
+			name:    "retrieving an added namespace returns the namespace",
+			addNs:   map[string]interface{}{"my-ns1": ns1, "my-ns2": ns2},
+			wantNs:  map[string]*corev1.Namespace{"my-ns1": ns1, "my-ns2": ns2},
+			wantErr: false,
+		},
+		{
+			name:    "adding a non-namespace type returns error",
+			addNs:   map[string]interface{}{"my-ns1": fooConstraint(), "my-ns2": ns2},
+			wantNs:  map[string]*corev1.Namespace{"my-ns2": ns2},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cache := newNsCache(nil)
+
+			for key, ns := range tt.addNs {
+				err := cache.Add(key, ns)
+				if !tt.wantErr && err != nil {
+					t.Errorf("unwanted error adding namespace")
+				}
+			}
+
+			for key, want := range tt.wantNs {
```

Comment:
This doesn't check for extraneous Namespaces which may be in cache. A way to ensure we don't have extra Namespaces (e.g. for the delete tests) would be to test the entire cache's map against `tt.wantNs`:

got := cache.cache
if diff := cmp.Diff(tt.wantNs, got); diff != "" {
  t.Errorf(diff)
}

Reply:
Good catch. I believe I achieved this by:
- verifying every namespace that should exist in the cache at the end of the test does, in fact, exist
- then verifying the cardinality of the cache equals the number of namespaces that should exist